test/e2e/federate.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	genericclient "github.com/kubernetes-sigs/federation-v2/pkg/client/generic"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefed2/federate"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/e2e/framework"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("Federate resource", func() {
+	f := framework.NewFederationFramework("federate-resource")
+	tl := framework.NewE2ELogger()
+	typeConfigFixtures := common.TypeConfigFixturesOrDie(tl)
+
+	var kubeConfig *restclient.Config
+	var client genericclient.Client
+	var typeConfig typeconfig.Interface
+	var testResourceName util.QualifiedName
+
+	BeforeEach(func() {
+		if kubeConfig == nil {
+			var err error
+			kubeConfig = f.KubeConfig()
+			client, err = genericclient.New(kubeConfig)
+			if err != nil {
+				tl.Fatalf("Error initializing dynamic client: %v", err)
+			}
+		}
+	})
+
+	AfterEach(func() {
+		if typeConfig != nil {
```

Comment:
Please try to minimize unnecessary indentation.  In cases like this one, I think the following would be preferable:

```golang
if typeConfig == nil {
  return
}
<deletion code>
```

I'm not sure why this check is keyed to `typeConfig`, since that doesn't seem indicate that any work has been done.  `testResourceName` seems like a better fit for determining whether the target resource needs to be deleted, and `federatedObject` for determining whether the federated resource needs to be deleted.  

Maybe, though, your intent is to attempt deletion regardless?  I'm not sure that's entirely safe with global variables, though.  A failure at any point in the test could result in a global variable retaining a value from a previous test.  Avoiding this possibility requires clearing global variables after cleanup (whether successful or unsuccessful).  I think this is reminding me why `defer()` might be preferable to `AfterEach`, since there is no reliance on global state.

Reply:
Very valid points. Updated to `defer()`!